## Predicting enhancers and promoters from human epigenomics data by spectral learning of Hidden Markov Models. ##
### Kevin Chen ###

- non-coding GWAS SNPs 
- [psychENCODE](http://www.nature.com/neuro/journal/v18/n12/full/nn.4156.html)
- experimental validation with iPSC and CRISPR

- spectral learning of HMMs for predicting chromatin states for states for a single cell
- extension to multiple cell types using HMMs with large structured data
- [spectacle](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0598-0)

- chromatin marks
  - H3K4me1 - histone 3 lysine 4 methyl

- chromatin state annotation via spectral learning
  - input:
  - output: 
- method of moments estimation, Pearson 1894
- express the expected moments of the data as functions of the parameters
- novel spectral learning approach
- tensor decomposition for learning HMMs

- [multi-view model](https://www2.eecs.berkeley.edu/Pubs/TechRpts/2014/EECS-2014-4.html)

- extension to multiple cell types
- [NIPS paper](https://arxiv.org/abs/1506.017440)

- other approaches 
  - [chromHMM](http://compbio.mit.edu/ChromHMM)
  - [ernst and kellis](http://www.nature.com/nbt/journal/v28/n8/abs/nbt.1662.html)
  - [variational EM approach](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-S5-S4)
